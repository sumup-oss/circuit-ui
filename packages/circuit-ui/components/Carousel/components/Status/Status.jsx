/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/react';

import Body from '../../../Body';
import { typography } from '../../../../styles/style-mixins';

const textStyles = ({ theme }) => css`
  ${theme.mq.untilKilo} {
    ${typography('two')(theme)};
  }
`;
const StyledText = styled(Body)(textStyles);

const Status = ({ step, total, ...props }) => (
  <StyledText variant="highlight" {...props}>
    {step + 1} / {total}
  </StyledText>
);

Status.propTypes = {
  /**
   * Current active index of a carousel.
   */
  step: PropTypes.number,
  /**
   * Total number of slides in a carousel.
   */
  total: PropTypes.number,
};

Status.defaultProps = {
  step: 0,
  total: 0,
};

export default Status;
