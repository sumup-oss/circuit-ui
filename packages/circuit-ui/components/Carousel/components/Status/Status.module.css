@media (max-width: 479px) {
  .base {
    font-size: var(--cui-typography-body-s-font-size);
    line-height: var(--cui-typography-body-s-line-height);
  }
}
