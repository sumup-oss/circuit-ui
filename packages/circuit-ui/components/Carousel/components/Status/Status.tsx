/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Body, type BodyProps } from '../../../Body/index.js';
import { clsx } from '../../../../styles/clsx.js';

import classes from './Status.module.css';

export interface StatusProps extends BodyProps {
  /**
   * Current active index of a carousel.
   */
  step: number;
  /**
   * Total number of slides in a carousel.
   */
  total: number;
}

export function Status({
  step = 0,
  total = 0,
  className,
  ...props
}: StatusProps) {
  return (
    <Body weight="bold" className={clsx(classes.base, className)} {...props}>
      {step + 1} / {total}
    </Body>
  );
}
