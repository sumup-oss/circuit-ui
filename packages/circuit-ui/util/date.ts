/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Temporal } from 'temporal-polyfill';
import { formatDateTime } from '@sumup-oss/intl';

import type { Locale } from './i18n.js';

export type FirstDayOfWeek = 1 | 7;
export type DaysInWeek = number;
export type PlainDateRange =
  | []
  | [Temporal.PlainDate]
  | [Temporal.PlainDate, Temporal.PlainDate];

// ISO 8601 timestamps only support positive 4-digit years
export const MIN_YEAR = 1;
export const MAX_YEAR = 9999;

export const MIN_MONTH = 1;
export const MAX_MONTH = 12;

export const MIN_DAY = 1;
// MAX_DAY is not constant as it depends on the year and month

export function getTodaysDate() {
  return Temporal.Now.plainDateISO();
}

export function isPlainDate(date: unknown): date is Temporal.PlainDate {
  return date instanceof Temporal.PlainDate;
}

export function toPlainDate(date?: string): Temporal.PlainDate | undefined {
  if (!date) {
    return undefined;
  }
  try {
    return Temporal.PlainDate.from(date);
  } catch {
    return undefined;
  }
}

export function sortDateRange<T extends PlainDateRange>(dateRange: T): T {
  return dateRange.sort((a, b) => Temporal.PlainDate.compare(a, b)) as T;
}

export function clampDate(
  date: Temporal.PlainDate,
  minDate?: Temporal.PlainDate | null,
  maxDate?: Temporal.PlainDate | null,
): Temporal.PlainDate {
  if (minDate && Temporal.PlainDate.compare(date, minDate) < 0) {
    return minDate;
  }
  if (maxDate && Temporal.PlainDate.compare(date, maxDate) > 0) {
    return maxDate;
  }
  return date;
}

export function getFirstDateOfWeek(
  date: Temporal.PlainDate,
  firstDayOfWeek: FirstDayOfWeek,
): Temporal.PlainDate {
  const diff =
    (date.dayOfWeek < firstDayOfWeek ? date.daysInWeek : 0) +
    date.dayOfWeek -
    firstDayOfWeek;
  return date.subtract({ days: diff });
}

export function getLastDateOfWeek(
  date: Temporal.PlainDate,
  firstDayOfWeek: FirstDayOfWeek,
): Temporal.PlainDate {
  return getFirstDateOfWeek(date, firstDayOfWeek).add({
    days: date.daysInWeek - 1,
  });
}

export function getMonthName(month: number, locale?: Locale) {
  // The year can be arbitrary since the month names are the same every year
  const yearMonth = new Temporal.PlainYearMonth(2000, month, 'gregory');
  return formatDateTime(yearMonth, locale, { month: 'long' });
}
