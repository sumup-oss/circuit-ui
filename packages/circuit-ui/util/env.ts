/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export function getEnvVariable(name: string) {
  if (name === 'NODE_ENV') {
    // Some bundlers have special logic for `process.env.NODE_ENV` which
    // relies on it being written as a continuous string. Destructuring or
    // dynamically accessing `NODE_ENV` can break this logic.
    throw new Error('Do not dynamically access NODE_ENV');
  }

  if (typeof process !== 'undefined') {
    return process.env?.[name];
  }

  return undefined;
}
