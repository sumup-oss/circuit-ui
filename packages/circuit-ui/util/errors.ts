/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable max-classes-per-file */
import React from 'react';

export class CircuitError extends Error {
  constructor(componentName: string, message: string) {
    super(`[${componentName}] ${message}`);
    this.name = 'CircuitError';
    // Adapted from https://stackoverflow.com/questions/33474179/react-access-parent-component-name

    /* eslint-disable max-len, no-underscore-dangle, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
    if (
      // @ts-expect-error Since this code only runs in development, it's fine to use this internal React API.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactCurrentOwner &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner
        .return &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner
        .return.type
    ) {
      this.stack =
        // @ts-expect-error This is fine.
        React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
          .ReactCurrentOwner.return.type as string;
    }
    if (
      // @ts-expect-error Since this code only runs in development, it's fine to use this internal React API.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactDebugCurrentFrame &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactDebugCurrentFrame.getCurrentStack
    ) {
      this.stack =
        // @ts-expect-error This is fine.
        React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactDebugCurrentFrame.getCurrentStack() as string;
    }
    /* eslint-enable max-len, no-underscore-dangle, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
  }
}

export class DeprecationError extends CircuitError {
  constructor(componentName: string, message: string) {
    super(componentName, message);
    this.name = 'DeprecationError';
  }
}

export class AccessibilityError extends CircuitError {
  constructor(componentName: string, message: string) {
    super(componentName, message);
    this.name = 'AccessibilityError';
  }
}
