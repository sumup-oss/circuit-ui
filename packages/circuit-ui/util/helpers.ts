/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isArray, isFunction, isObject, isString } from './type-check';

/**
 * Calls each function in an array with the arguments it receives.
 * Falsy functions are ignored.
 */
export function eachFn<Args extends []>(
  fns: (undefined | ((...args: Args) => unknown))[],
) {
  return (...args: Args): void =>
    fns.forEach((fn) => isFunction(fn) && fn(...args));
}

/**
 * Checks whether an array, object, or string is empty or whether the value is
 * falsy.
 */
export function isEmpty(value: unknown): boolean {
  if (!value) {
    return true;
  }

  if (isString(value) || isArray(value)) {
    return !value.length;
  }

  if (isObject(value)) {
    return !Object.keys(value).length;
  }

  return false;
}

export function clamp(value: number, min: number, max: number): number {
  if (process.env.NODE_ENV !== 'production' && min >= max) {
    throw new RangeError(
      `The minimum value (${min}) must be less than the maximum value (${max}).`,
    );
  }
  return Math.max(min, Math.min(value, max));
}

type Fn<T extends []> = (...args: T) => void;

/**
 * Triggers a function at most once in a given amount of time.
 */
export function throttle<T extends []>(fn: Fn<T>, timeout: number): Fn<T> {
  let ready = true;
  return (...args: T) => {
    if (!ready) {
      return;
    }

    ready = false;
    fn(...args);
    setTimeout(() => {
      ready = true;
    }, timeout);
  };
}
