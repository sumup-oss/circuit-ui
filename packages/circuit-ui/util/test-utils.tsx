/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FunctionComponent, ReactElement } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, RenderResult } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import {
  ComponentsContext,
  defaultComponents,
} from '../components/ComponentsContext/ComponentsContext';

export * from '@testing-library/react';

export type RenderFn<T = any> = (component: ReactElement, ...rest: any) => T;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const render: RenderFn<RenderResult> = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn<string> = (component) =>
  renderToStaticMarkup(<WithProviders>{component}</WithProviders>);
const create = (...args: Parameters<RenderFn<RenderResult>>) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export { create, render, renderToHtml, renderHook, actHook, userEvent, axe };
