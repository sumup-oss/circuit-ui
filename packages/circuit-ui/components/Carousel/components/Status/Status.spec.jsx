/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { axe, render } from '../../../../util/test-utils';

import Status from './Status';

describe('Status', () => {
  describe('styles', () => {
    it('should render with default styles', () => {
      const { container } = render(<Status step={1} total={3} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const { container } = render(<Status step={1} total={3} />);
      const actual = await axe(container);

      expect(actual).toHaveNoViolations();
    });
  });
});
