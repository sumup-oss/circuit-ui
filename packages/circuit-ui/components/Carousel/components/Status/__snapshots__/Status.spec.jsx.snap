// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Status > styles > should render with default styles 1`] = `
.circuit-0 {
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 700;
}

@media (max-width: 479px) {
  .circuit-0 {
    font-size: 0.875rem;
    line-height: 1.25rem;
  }
}

<div>
  <strong
    class="circuit-0"
  >
    2
     / 
    3
  </strong>
</div>
`;
